### PR TITLE
Flexible restart write times (restart_fh) for med, ocn, ice, wav

### PR DESCRIPTION
**Pull request by NickSzapiro-NOAA**
## Commit Queue Requirements:

- [x] Fill out all sections of this template.
- [x] All sub component pull requests have been reviewed by their code managers.
- [x] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
- [x] Commit 'test_changes.list' from previous step
---
## Description:

This PR enables writing forecast hour defined restarts ("restart_fh") to have option of forecast hour restarts in addition to existing functionality like RESTART_N. Restart writes are triggered when input restart_fh forecast hours are evenly divisible by a component's timestep (internally compared in units of integer seconds) and skipped otherwise. The implementation builds on [current restart_fh in MOM6](https://github.com/ufs-community/ufs-weather-model/pull/2099) for CMEPS, MOM6, CICE, and WW3 via a new ``shr_is_restart_fh_mod`` shared module (located at UFS level in CDEPS-interface). With this, restart writes can be triggered at any forecast time via UFS configuration for these components.

Currently, space-delimited floating point forecast hours are input in model_configure attributes, like `restart_fh: 0.25 2.5 6 17 24`. Such an attribute is well tested (e.g., output_fh for FV3 implemented in https://github.com/NOAA-EMC/fv3atm/pull/345), but it may be more natural to move attribute to ufs.configure for all components once issue with reading list of floating points is resolved (https://github.com/orgs/esmf-org/discussions/291).

Testing with cpld_control_gfsv17 confirms that restarts are bit-for-bit when (1) sharing common times with RESTART_N and (2) instead of RESTART_N. This option is exercised in the cpld_control_p8 RT for med, ocn, ice, wav with restart files added to the baselines for comparison. Baselines are otherwise unchanged.

### Commit Message:

```
* UFSWM - Flexible restart write times (restart_fh) for med, ocn, ice, wav
  * CICE - Flexible restart write times (restart_fh)
  * CMEPS - Flexible restart write times (restart_fh)
  * MOM6 - Flexible restart write times (restart_fh)
  * WW3 - Flexible restart write times (restart_fh)
```

### Priority:

* Normal

### UFSWM:

* Closes # https://github.com/ufs-community/ufs-weather-model/issues/2348 . Note that FV3 is separate with its own encoded restart_interval implementation

### Sub component Pull Requests:

* CICE: https://github.com/NOAA-EMC/CICE/pull/85
* CMEPS: https://github.com/NOAA-EMC/CMEPS/pull/125
* MOM6: https://github.com/NOAA-EMC/MOM6/pull/139
* WW3: https://github.com/NOAA-EMC/WW3/pull/1320

### UFSWM Blocking Dependencies:

* https://github.com/ufs-community/ufs-weather-model/pull/2445
* https://github.com/ufs-community/ufs-weather-model/issues/2502

---
### Regression Test Changes (Please commit test_changes.list):

* Adds additional restart files to baseline

### Input data Changes:

* None

### Library Changes/Upgrades:

* No Updates
  
---

## Testing Log:
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Hercules
  - [x] Jet
  - [x] Gaea
  - [x] Derecho
- WCOSS2
  - [x] Dogwood/Cactus
  - [x] Acorn
- [ ] CI
- [ ] opnReqTest (complete task if unnecessary)

### CODE REVIEW

**Comment by DeniseWorthen** [comment]
@NickSzapiro-NOAA Could you start making the required changes in WW3 also, starting with my PR branch for the netCDF restarts? That branch is https://github.com/NOAA-EMC/WW3/pull/1303. That would allow all components to have this feature available at the same time, which I think would be a priority. 

There are script level changes required in UWM also to enable PIO in WW3. See my UWM PR https://github.com/ufs-community/ufs-weather-model/pull/2445.

**Comment by NickSzapiro-NOAA** [comment]
Sounds great, @DeniseWorthen! 

So the "restart_fh" module is now shared at:
https://github.com/NickSzapiro-NOAA/ufs-weather-model/blob/restart_fh_marine/CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90

For WW3, it looks like it makes sense to keep restart_fh near alarm_restart in model/src/wav_comp_nuopc.F90.
If we init in ModelSetRunClock and check in ModelAdvance, this is the same as in MOM6 and CICE.

May I check that there is no concern about [conflicts between the restart strides](https://github.com/NOAA-EMC/WW3/issues/216) with the netcdf output option?

**Comment by DeniseWorthen** [comment]
@NickSzapiro-NOAA Thanks. My gut instinct is that there will be no similar stride conflict. When ``use_restartnc`` is set, the code totally bypasses the current code which determines whether a restart will be written. See for example the use of the logical flag (and it's .not. variant) in w3wavemd.F90 in my feature branch.

File: CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90 [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    character(len=*), parameter :: subname = 'init_restart_fh'
+    !-----------------------------------------------------------------------
+
+    ! set up Times to write non-interval restarts
+    inquire(FILE='model_configure', EXIST=isPresent)
+    if (isPresent) then !model_configure exists. this is ufs run
+      CF_mc = ESMF_ConfigCreate(rc=rc)
+      call ESMF_ConfigLoadFile(config=CF_mc,filename='model_configure' ,rc=rc)
+      !if (ChkErr(rc,__LINE__,__FILE__)) return
+      if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
```
Comment (DeniseWorthen) [inline CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90:53]: Why can't you use the cleaner form ? I see it commented out.
```
if (ChkErr(rc,__LINE__,u_FILE_u)) return
```
Add

use dshr_methods_mod , only : chkerr
```
character(*), parameter :: u_FILE_u = &
       __FILE__
```

Comment (NickSzapiro-NOAA) [reply CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90:53]: We can, but that brings additional dependencies into this module besides ESMF. There are also quite some duplicate ChkErr functions floating around all just wrapping an ESMF function

File: CICE-interface/CMakeLists.txt [hunk 2]
```diff
@@ -63,7 +63,7 @@ list(APPEND lib_src_files
   ${icepack_files}
   ${cice_mpi_comm_files}
   ${cice_nuopc_cmeps_driver_files}
-  ${cice_cdeps_inline_files})
+  ${cice_cdeps_share_files})
```
Comment (junwang-noaa) [inline CICE-interface/CMakeLists.txt:66]: I know this is a temporary solution. Does this mean that now cice/mom/ww3 caps now have dependency on the CDEPS share code?

Comment (NickSzapiro-NOAA) [reply CICE-interface/CMakeLists.txt:66]: Yes. 

CICE and CMEPS already have dependency for CDEPS inline functionality.

MOM6 and WW3 now have restart_fh module in cmakes. While this is currently under CDEPS-interface, it is a ufs-weather-model file (not a CDEPS subcomponent file)

**Comment by FernandoAndrade-NOAA** [comment]
#2483 has been merged, which included #2512 to fix the #2502 blocker, please go ahead and sync / resolve conflicts.

**Comment by NickSzapiro-NOAA** [comment]
@FernandoAndrade-NOAA Resolved conflicts. Should be ready

**Comment by NickSzapiro-NOAA** [comment]
I know I've mucked this up before, but I don't think a new bl_date is needed. 
Only these new files are added to the cpld_control_p8 baseline:
```
RESTART/20210322.090000.MOM.res.nc
RESTART/iced.2021-03-22-32400.nc 
RESTART/ufs.cpld.cpl.r.2021-03-22-32400.nc 
ufs.cpld.ww3.r.2021-03-22-32400.nc
```

**Comment by FernandoAndrade-NOAA** [comment]
Ok, we should be all set with the testing, I'll let sub components know to continue with merging updates.

File: CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90 (hunk 1 above)
Comment (FernandoAndrade-NOAA) [reply CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90:53]: Marking as resolved, please let me know if this requires a new separate issue.

**Comment by jiandewang** [comment]
> Ok, we should be all set with the testing, I'll let sub components know to continue with merging updates.

time for me to do the MOM6 merging, right ?

**Comment by FernandoAndrade-NOAA** [comment]
> > Ok, we should be all set with the testing, I'll let sub components know to continue with merging updates.
> 
> time for me to do the MOM6 merging, right ?

Correct, please continue with merging on the sub component side.

**Comment by jiandewang** [comment]
MOM6 merged, hash # 8cc8b70f

**Comment by NickSzapiro-NOAA** [comment]
Thank you, @jiandewang . CICE merged, hash # ff7fd76

**Comment by JessicaMeixner-NOAA** [comment]
WW3 hash is 29063ec7b631d3a967172e8c45c5af13a18e0a82

**Review by FernandoAndrade-NOAA — APPROVED** [review]

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Pull request merged**